### PR TITLE
feat(client): Add signin authorization support.

### LINKED ISSUE #210: Add signin-authorization endpoint support
Proposed new endpoints:

* /account/login/send_authorization_code
* /account/login/verify_authorization_code
* /account/login/check_authorization
* /account/login/reject_authorization_code

### PR DESCRIPTION
**Pull request by shane-tomlinson**
fixes #210

### CODE REVIEW

File: client/FxAccountClient.js [hunk 1]
```diff
@@ -1060,6 +1060,86 @@ define([
       });
   };
 
+  /**
+   * Send an login authorization code
```
Comment (vladikoff) [inline client/FxAccountClient.js]: `an login` - > ` a login` ?

File: client/FxAccountClient.js [hunk 2]
```diff
@@ -1060,6 +1060,86 @@ define([
...
+        data.resume = options.resume;
+      }
+    }
+
+    var request = this.request;
+    return request.send('/account/login/send_authorization_code', 'POST', null, data);
+  };
+
+  /**
+   * @method rejectLoginAuthorizationCode
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: Add a description.

File: client/FxAccountClient.js [hunk 3]
```diff
@@ -220,6 +222,10 @@ define([
             data.metricsContext = metricsContext.marshall(options.metricsContext);
           }
 
+          if (options.code) {
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: rename to `unblockCode` for consistency?

File: client/FxAccountClient.js [hunk 4]
```diff
@@ -1060,6 +1066,45 @@ define([
...
+   */
+  FxAccountClient.prototype.sendUnblockCode = function (email) {
+    required(email, 'email');
+
+    var data = {
+      email: email
+    };
+
+    var request = this.request;
+    return request.send('/account/login/send_authorization_code', 'POST', null, data);
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: `authorization` => `unblock`

File: client/FxAccountClient.js [hunk 5]
```diff
@@ -1060,6 +1066,45 @@ define([
...
+    required(uid, 'uid');
+    required(code, 'code');
+
+    var data = {
+      code: code,
+      uid: uid
+    };
+
+    var request = this.request;
+    return request.send('/account/login/reject_authorization_code', 'POST', null, data);
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: `authorization` => `unblock`

File: client/FxAccountClient.js [hunk 6]
```diff
@@ -1060,6 +1066,45 @@ define([
...
+   * @param {String} uid Account ID
+   * @param {String} code Verification code
+   * @return {Promise} A promise that will be fulfilled with JSON `xhr.responseText` of the request
+   */
+  FxAccountClient.prototype.rejectUnblockCode = function (uid, code) {
+    required(uid, 'uid');
+    required(code, 'code');
+
+    var data = {
+      code: code,
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: `unblockCode`?

**Comment by shane-tomlinson** [comment]
@seanmonstar - could you start to review this one too? The tests won't pass until we have a working auth-server, so we shouldn't merge until then, still want to get yoru feedback on whether this is the API you expect.

File: client/FxAccountClient.js [hunk 7]
```diff
@@ -193,6 +193,8 @@ define([
    *     @param {String} [options.metricsContext.utmMedium] marketing campaign medium
    *     @param {String} [options.metricsContext.utmSource] marketing campaign source
    *     @param {String} [options.metricsContext.utmTerm] marketing campaign search term
+   *   @param {String} [options.unblockCode]
```
Comment (seanmonstar) [inline client/FxAccountClient.js]: Indentation seems off.

Comment (shane-tomlinson) [inline client/FxAccountClient.js]: That's cause the params above it have extra indentation to handle being 2 levels deep in the `options`.

**Comment by shane-tomlinson** [comment]
Blocked on auth-server support for the feature and a deploy box to test against that sends all `block.*@restmail.net` users through the unblock flow.

**Comment by shane-tomlinson** [comment]
@vladikoff - You are our resident js-client expert. Can I get your r on this too to make sure I have all the parts correct? The `npm run test-local` functional tests won't pass until the auth server supports the endpoints, so we shouldn't merge until that's ready.

File: client/FxAccountClient.js [hunk 8]
```diff
@@ -1071,6 +1077,45 @@ define([
...
+   * @return {Promise} A promise that will be fulfilled with JSON `xhr.responseText` of the request
+   */
+  FxAccountClient.prototype.sendUnblockCode = function (email) {
+    required(email, 'email');
+
+    var data = {
+      email: email
+    };
+
+    var request = this.request;
```
Comment (vladikoff) [inline client/FxAccountClient.js]: do you need this or can you just call `this.request` below?

File: tests/addons/restmail.js [hunk 9]
```diff
@@ -21,7 +21,7 @@ define([
 
     return self.request.send('/mail/' + user, 'GET')
       .then(function (result) {
-        if (result.length === number) {
+        if (result.length >= number) {
```
Comment (vladikoff) [inline tests/addons/restmail.js]: scary!

File: client/FxAccountClient.js (hunk 8 above)
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: Ya, good point.

File: tests/addons/restmail.js (hunk 9 above)
Comment (shane-tomlinson) [inline tests/addons/restmail.js]: Yeah, there was some strangeness with the mock emails that I didn't know about - if you have too many mock emails result.length > number and the tests timeout.

File: client/FxAccountClient.js (hunk 8 above)
Comment (seanmonstar) [inline client/FxAccountClient.js]: I noticed this also, but then expanding the file, I notice this pattern in many of the methods, so felt it was just for consistency. Should they all be changed (different PR)?

Comment (shane-tomlinson) [inline client/FxAccountClient.js]: I just audited the rest of the file, everywhere there is a `var request = this.request`, `request` is used inside of a callback, so there was a real reason to set the reference. I had no such reason.

**Pull request merged**